hera_cal/redcal.py

```diff
@@ -789,3 +837,335 @@ def is_redundantly_calibratable(antpos, bl_error_tol=1.0):
+        append_to_history: string to add to history of output firstcal and omnical files
+        verbose: print calibration progress updates
+
+    Returns:
+        cal: the dictionary result of the final run of redcal_partial_io_iteration (see above for details)
+    '''
+    assert isinstance(input_data, str), 'input_data must be a single string path to a uvh5 visibility data file.'
+    hd = HERAData(input_data)
+    ex_ants = set(ex_ants)
+    if ant_metrics_file is not None:
+        for ant in load_metric_file(ant_metrics_file)['xants']:
+            ex_ants.add(ant)
+
+    # loop over calibration, removing bad antennas and re-running if necessary
+    while True:
+        # Run redundant calibration
+        if verbose:
+            print('\nNow running redundant calibration without antennas', list(ex_ants), '...')
+        cal = redcal_partial_io_iteration(hd, nInt_to_load=nInt_to_load, pol_mode=pol_mode, ex_ants=ex_ants, 
+                                          solar_horizon=solar_horizon, conv_crit=conv_crit, maxiter=maxiter, 
+                                          check_every=check_every, check_after=check_after, gain=gain, verbose=verbose)
+        
+        # Determine whether to add additional antennas to exclude
+        z_scores = per_antenna_modified_z_scores({ant: np.nanmedian(cspa) for ant, cspa in cal['chisq_per_ant'].items() 
+                                                  if not np.all(cspa == 0)})
+        n_ex = len(ex_ants)
+        for ant, score in z_scores.items():
+            if (score >= ant_z_thresh):
+                ex_ants.add(ant[0])
+                if verbose:
+                    print('Throwing out antenna', ant[0], 'for a z-score', score, 'on polarization', ant[1])
+        if len(ex_ants) == n_ex:
+            break
+
+    # output results files
+    if outdir is None:
+        outdir = os.path.dirname(input_data)
+
+    if verbose:
+        print('\nNow saving firstcal gains to', input_data + firstcal_suffix)
+    write_cal(os.path.basename(input_data) + firstcal_suffix, cal['g_firstcal'], hd.freqs, hd.times, 
+              flags=cal['gf_firstcal'], outdir=outdir, overwrite=clobber, history=append_to_history)
+    
+    if verbose:
+        print('Now saving omnical gains to', input_data + omnical_suffix)
+    write_cal(os.path.basename(input_data) + omnical_suffix, cal['g_omnical'], hd.freqs, hd.times, 
+              flags=cal['gf_omnical'], quality=cal['chisq_per_ant'], total_qual=cal['chisq'], 
+              outdir=outdir, overwrite=clobber, history=append_to_history)
+
+    if verbose:
+        print('Now saving omnical visibilities to', input_data + omnivis_suffix)
+    hd.read(bls=cal['v_omnical'].keys())
+    hd.update(data=cal['v_omnical'], flags=cal['vf_omnical'])
+    hd.history += append_to_history
+    hd.write_uvh5(os.path.join(outdir, os.path.basename(input_data) + omnivis_suffix), clobber=True)
+
+    return cal
+
+
+def redcal_argparser():
```

Comment:
We are bordering on a level where we might want to move to config files to control parameter passing to these scripts.

Reply:
If we do that, it should be repo-wide. And in practice, using hera_opm, we ultimately do that. For now, I think we're fine.

Reply:
Moved to issue #378